From a2f74b141a01b30360ecc20271377664acb16841 Mon Sep 17 00:00:00 2001
From: Hridya Valsaraju <hridya@google.com>
Date: Fri, 27 Jul 2018 22:09:12 -0700
Subject: [PATCH 48/49] recovery: Support configfs usb configuration

Backported from upstream I7891bb84427ec734a21a872036629b95ab3fb13c
"Add fastboot mode to recovery"

Change-Id: I252f99f0e44f6be53c5b7577c19c8c1f0f616ad8
---
 etc/init.rc  | 61 +++++++++++++++++++++++++++++++++++++++-------------
 recovery.cpp |  1 +
 2 files changed, 47 insertions(+), 15 deletions(-)

diff --git a/etc/init.rc b/etc/init.rc
index 3775e375..8449f6f7 100644
--- a/etc/init.rc
+++ b/etc/init.rc
@@ -6,6 +6,8 @@ on early-init
 
     start ueventd
 
+    setprop sys.usb.configfs 0
+
 on init
     export ANDROID_ROOT /system
     export ANDROID_DATA /data
@@ -44,20 +46,6 @@ on init
     # pstore/ramoops previous console log
     mount pstore pstore /sys/fs/pstore
 
-on fs
-    write /sys/class/android_usb/android0/f_ffs/aliases adb
-    mkdir /dev/usb-ffs 0770 shell shell
-    mkdir /dev/usb-ffs/adb 0770 shell shell
-    mount functionfs adb /dev/usb-ffs/adb uid=2000,gid=2000
-
-    write /sys/class/android_usb/android0/enable 0
-    write /sys/class/android_usb/android0/idVendor 18D1
-    write /sys/class/android_usb/android0/idProduct D001
-    write /sys/class/android_usb/android0/functions adb
-    write /sys/class/android_usb/android0/iManufacturer ${ro.product.manufacturer}
-    write /sys/class/android_usb/android0/iProduct ${ro.product.model}
-    write /sys/class/android_usb/android0/iSerial ${ro.serialno}
-
 on boot
     ifup lo
     hostname localhost
@@ -114,6 +102,49 @@ service adbd /sbin/adbd --root_seclabel=u:r:su:s0 --device_banner=recovery
 
 # Restart adbd so it can run as root
 on property:lineage.service.adb.root=1
-    write /sys/class/android_usb/android0/enable 0
     restart adbd
+
+# setup_adbd will start adb once it has checked the keys
+on fs && property:ro.debuggable=1
+    start setup_adbd
+
+on fs && property:sys.usb.configfs=1
+    mount configfs none /config
+    mkdir /config/usb_gadget/g1 0770 shell shell
+    write /config/usb_gadget/g1/idVendor 0x18D1
+    mkdir /config/usb_gadget/g1/strings/0x409 0770
+    write /config/usb_gadget/g1/strings/0x409/serialnumber ${ro.serialno}
+    write /config/usb_gadget/g1/strings/0x409/manufacturer ${ro.product.manufacturer}
+    write /config/usb_gadget/g1/strings/0x409/product ${ro.product.model}
+    mkdir /config/usb_gadget/g1/functions/ffs.adb
+    mkdir /config/usb_gadget/g1/configs/b.1 0777 shell shell
+    mkdir /config/usb_gadget/g1/configs/b.1/strings/0x409 0770 shell shell
+
+on fs && property:sys.usb.configfs=0
+    write /sys/class/android_usb/android0/f_ffs/aliases adb
+    write /sys/class/android_usb/android0/idVendor 18D1
+    write /sys/class/android_usb/android0/iManufacturer ${ro.product.manufacturer}
+    write /sys/class/android_usb/android0/iProduct ${ro.product.model}
+    write /sys/class/android_usb/android0/iSerial ${ro.serialno}
+
+on fs
+    mkdir /dev/usb-ffs 0775 shell shell
+    mkdir /dev/usb-ffs/adb 0770 shell shell
+    mount functionfs adb /dev/usb-ffs/adb uid=2000,gid=2000
+
+on property:sys.usb.config=adb
+    start adbd
+
+on property:sys.usb.config=adb && property:sys.usb.configfs=0
+    write /sys/class/android_usb/android0/idProduct D001
+    write /sys/class/android_usb/android0/functions adb
     write /sys/class/android_usb/android0/enable 1
+    setprop sys.usb.state ${sys.usb.config}
+
+# Configfs trigger
+on property:sys.usb.config=adb && property:sys.usb.ffs.ready=1 && property:sys.usb.configfs=1
+    write /config/usb_gadget/g1/idProduct 0xD001
+    write /config/usb_gadget/g1/configs/b.1/strings/0x409/configuration "adb"
+    symlink /config/usb_gadget/g1/functions/ffs.adb /config/usb_gadget/g1/configs/b.1/f1
+    write /config/usb_gadget/g1/UDC ${sys.usb.controller}
+    setprop sys.usb.state ${sys.usb.config}
diff --git a/recovery.cpp b/recovery.cpp
index ec6697fa..ac200578 100644
--- a/recovery.cpp
+++ b/recovery.cpp
@@ -1645,6 +1645,7 @@ static void setup_adbd() {
   }
 
   // Trigger (re)start of adb daemon
+  property_set("sys.usb.config", "adb");
   property_set("lineage.service.adb.root", "1");
 }
 
-- 
2.17.1

